compiler/rustc_span/src/hygiene.rs

```diff
@@ -571,6 +605,21 @@ pub fn walk_chain(span: Span, to: SyntaxContext) -> Span {
     HygieneData::with(|data| data.walk_chain(span, to))
 }
 
+pub fn walk_chain_collapsed(
+    span: Span,
+    to: SyntaxContext,
+    collapse_enabled: bool,
+    attr_enabled: bool,
+) -> Span {
+    HygieneData::with(|hdata| {
+        if collapse_enabled && span.from_expansion() {
```

Comment:
The `from_expansion` condition is redundant here, I think.
If the span is not from an expansion, then the loop in `HygieneData::walk_chain_collapsed` will give the same result.